teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md

```diff
@@ -0,0 +1,502 @@
+```
+{
+        "uri": "<URI>",
+        "digest": { /* DigestSet */ }
+ }
+```
+The other specific requirements of these fields can be found in [in-toto attestation spec](https://github.com/in-toto/attestation/tree/main/spec). Another thing to be aware is that For `digest`, it has a key for the digest type and has a value for that key. Git, Mercurial, Bazaar and Subversion have full VCS location support. CVS, Perforce are providing this VCS location as plain URLs. For VCS, the list of VCS that the community are looking into is: Git, Mercurial, Subversion, Bazaar, CVS, Perforce. Currently Chains only supports Git revisions for VCS and OCI images as signable targets for inputs.
+
+For the `outputs`, we are looking at the following support, as these are the artifact types supported in Artifact Registries. However, the list should definitely be expanded as this doc is spreading the discussions in the communities to see which should be prioritized. Currently, we have the following under considerations: Python, Maven, Go, NodeJS and OCI images. Note that OCI images are the ones we currently support.
+
+#### Concrete Use Cases
+All below examples will be generating in-toto attestation.
+ 1. Use Git commits as sources, and generates in-toto provenance for a TaskRun that builds an image.
+   
+    A Task example `without` structured results and params support can be like:
+    ``` yaml
+    apiVersion: tekton.dev/v1beta1
+    kind: Task
+    spec:
+      params:
+        - name: CHAINS-GIT_COMMIT
+          type: string
+          description: git commit sha
+        - name: CHAINS-GIT_URL
+          type: string
+          description: git commit url
+        - name: FOO_IMAGE
+          type: string
+        - description: artifact image
+      ...
+      results:
+        - name: IMAGE_DIGEST
+          description: digest of image
+        - name: IMAGE_URL
+          description: url of image
+    ```
+    A Task example `with` structured results support will be like:
+    ``` yaml
+    apiVersion: tekton.dev/v1beta1
+    kind: Task
+    spec:
+      ...
+      results:
+        - name: ARTIFACT_INPUTS
+          type: array
+          description: |
+            Stores result names that should be captured as signable targets as "materials" in in-toto provenances in Tekton Chains.
+        - name: ARTIFACT_OUTPUTs
+          type: array
+          description: |
+            Stores result names that should be captured as signable targets as "subjects" in in-toto provenances in Tekton Chains.
+        - name: git-vcs
+          type: object
+          description: |
+            The source distribution
+            * uri: resource uri of the artifact.
+            * digest: revision digest in form of algorithm:digest.
+            * artifact_type: artifact type categorizes the artifacts.
```

Comment:
I added more explanations to `artifact_type` in the article. Currently I'm more using this for Chains to monitor metrics. I will mention this to SLSA folks. I also created an issue in Intoto provenance repo: https://github.com/in-toto/attestation/issues/103

Reply:
Updates from https://github.com/in-toto/attestation/issues/103 - pipeline authors can follow PACKAGE URL format to provide the URI, and this way it can identify the artifact type. However, consider the situation that pipeline authors can't follow the PACKAGE URL format and still want to identify the artifact type, and this `artifact_type ` field will become useful. But I would remove this field, since there is no essential reason to add the field here, and people can still provide URI in PACKAGE URL to identify the artifact type.